lib/sly-completion.el

```diff
@@ -216,9 +216,28 @@ COMPLETIONS is a list of propertized strings."
            finally return (list formatted nil)))
 
 (defun sly-completion-annotation (completion)
-  "Grab the annotation of COMPLETION, a string, if any"
+  "Grab the annotation of COMPLETION, a string, if any."
   (get-text-property 0 'sly--annotation completion))
 
+(defun sly--completion-company-kind (completion)
+  "Convert the COMPLETION annotation to a company-kind supported symbol."
+  (let ((annot
+         (sly-completion-annotation completion)))
+    (when-let* ((comps (split-string annot " "))
```

Comment:
This `split-string` is a code smell, right?  :-) Why should we be reverse-engineering the annotation which contains classification + score, when we are in control of the code and can just store them separately?

Reply:
Here's how you should do this:

```lisp
diff --git a/lib/sly-completion.el b/lib/sly-completion.el
index 5839ed45..abd16942 100644
--- a/lib/sly-completion.el
+++ b/lib/sly-completion.el
@@ -204,10 +204,8 @@ COMPLETIONS is a list of propertized strings."
                                                completion))
            (add-text-properties 0
                                 (length completion)
-                                `(sly--annotation
-                                  ,(format "%s %5.2f%%"
-                                           classification
-                                           (* score 100))
+                                `(sly--classification ,classification
+                                  sly--score ,score
                                   sly--suggestion
                                   ,suggestion)
                                 completion)
@@ -217,7 +215,9 @@ COMPLETIONS is a list of propertized strings."
 
 (defun sly-completion-annotation (completion)
   "Grab the annotation of COMPLETION, a string, if any"
-  (get-text-property 0 'sly--annotation completion))
+  (format "%s %5.2f%%"
+          (get-text-property 0 'sly--classification completion)
+          (* (get-text-property 0 'sly--score completion) 100)))
 
 ;;; backward-compatibility
 (defun sly-fuzzy-completions (pattern)
@@ -323,6 +323,10 @@ ANNOTATION) describing each completion possibility."
                 (when (buffer-live-p buffer) ; on the safe side
                   (cons buffer (with-current-buffer buffer
                                  (point)))))))
+          :company-kind
+          (lambda (obj)
+            ;; do the mapping thing
+            )
           :company-prefix-length
           (and (sly--completion-inside-string-or-comment-p) 0))))
 
@@ -601,8 +605,7 @@ Intended to go into `completion-at-point-functions'"
              sly--completion-explanation))
     (cl-loop with first = (point)
              for completion in completions
-             for annotation = (or (get-text-property 0 'sly--annotation completion)
-                                  "")
+             for annotation = (or (sly-completion-annotation completion) "")
              for start = (point)
              do
              (cl-loop for (beg . end) in
```

Reply:
Code smell, yeah. Hacked this together in 15 minutes and was surprised that I got something working. :)
I will have a look and will change it to what you suggested.